# Test Run

In [2]:
import requests

In [79]:
r = requests.get('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1955/')

In [5]:
r.status_code

200

In [6]:
r.status_code == requests.codes.ok

True

In [7]:
r.text

'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">\n<html>\n<head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">\n<title>\n      Fortune 500 Companies - Archived List of Best Companies from 1995\n    </title>\n<meta name="Description" content="FORTUNE Magazine\'s annual ranking of America\'s 500 largest corporations">\n<meta name="keywords" content="FORTUNE, FORTUNE magazine, finance, FORTUNE home, FORTUNE 500, Fortune.com, best companies, top employers, best employers, business research, business and finance, business profiles, fortune list, business list, America\'s biggest companies, 500 list">\n<meta http-equiv="Pragma" content="no-cache">\n<script language="JavaScript1.1" src="http://i.cnn.net/money/.element/ssi/javascript/1.0/main.js" type="text/javascript">\n</script><link rel="stylesheet" href="http://i.cnn.net/money/.element/ssi/css/1.0/main.css" type="text/css">\n<link rel="stylesheet" href="http://i.cnn.net/money/.element/ssi/css/1.1/mag_l

In [80]:
from bs4 import BeautifulSoup

In [81]:
soup = BeautifulSoup(r.text)

In [13]:
print(soup.prettify)

<bound method Tag.prettify of <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<title>
      Fortune 500 Companies - Archived List of Best Companies from 1995
    </title>
<meta content="FORTUNE Magazine's annual ranking of America's 500 largest corporations" name="Description"/>
<meta content="FORTUNE, FORTUNE magazine, finance, FORTUNE home, FORTUNE 500, Fortune.com, best companies, top employers, best employers, business research, business and finance, business profiles, fortune list, business list, America's biggest companies, 500 list" name="keywords"/>
<meta content="no-cache" http-equiv="Pragma"/>
<script language="JavaScript1.1" src="http://i.cnn.net/money/.element/ssi/javascript/1.0/main.js" type="text/javascript">
</script><link href="http://i.cnn.net/money/.element/ssi/css/1.0/main.css" rel="stylesheet" type="text/css"/>
<link href="http://i.cnn.net/money/.element/ssi/css/1.1/mag

In [29]:
a = soup.find_all('a')

In [39]:
type(a)

bs4.element.ResultSet

In [35]:
soup.find(href = "/assets/money.cnn.com/snapshots/1955/563.html")

<a href="/assets/money.cnn.com/snapshots/1955/563.html">General Motors</a>

In [82]:
entries = []
for wrapper in soup.find_all('a'):
   entries.append(wrapper.text)

In [83]:
entries

['',
 'Companies',
 'Profits',
 'Assets',
 'Current FORTUNE 500',
 'General Motors',
 'Exxon Mobil',
 'U.S. Steel',
 'General Electric',
 'Esmark',
 'Chrysler',
 'Armour',
 'Gulf Oil',
 'Mobil',
 'DuPont',
 'Amoco',
 'Bethlehem Steel',
 'CBS',
 'Texaco',
 'AT&T Technologies',
 'Shell Oil',
 'Kraft',
 'ChevronTexaco',
 'Goodyear Tire & Rubber',
 'Boeing',
 'Sinclair Oil',
 'Navistar International',
 'RCA',
 'Union Carbide',
 'Firestone Tire & Rubber',
 'Douglas Aircraft',
 'Procter & Gamble',
 'Republic Steel',
 'Citgo Petroleum',
 'ConocoPhillips',
 'General Foods',
 'Uniroyal',
 'Borden Chemical',
 'Lockheed Martin',
 'Alcoa',
 'International Paper',
 'Wilson',
 'Sunoco',
 'United Technologies',
 'American Can',
 'General Dynamics',
 'Rockwell Automation',
 'Eastman Kodak',
 'Goodrich',
 'Continental Group',
 'Bendix',
 'Atlantic Richfield',
 'Fortune Brands',
 'Ryerson Tull',
 'ARMCO',
 'Honeywell Intl.',
 'AMAX',
 'Conoco',
 'Allis-Chalmers',
 'Jones & Laughlin Steel',
 'General Mil

In [84]:
companies = entries[5:-11]

In [64]:
len(companies)

100

In [59]:
import pandas as pd

In [101]:
F100 = pd.DataFrame(data = companies, columns = ['Company Name'])

# Pipelines

## Imports

In [63]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

## Link Pipeline - Manual

In [104]:
def update_list(df, url):
    r = requests.get(url) # gets all the data from a webpage
    assert r.status_code == requests.codes.ok # checks that we got something
    soup = BeautifulSoup(r.text)
    
    entries = [] # list of topics from the webpage, includes some extraneous info
    for wrapper in soup.find_all('a'):
        entries.append(wrapper.text)
    
    companies = entries[5:-11] # slices the irrelevant information
    assert len(companies) == 100 # should only have 100 companies
    
    F5 = pd.DataFrame(data = companies, columns = df.columns) # create a dataframe from the list to concatenate
    assert df.columns == F5.columns # should be identical because reference above
    assert df.shape[1] == F5.shape[1] # verifies they have the same number of columns
    
    df = pd.concat([df, F5], 
                   ignore_index=True) # ignore index to keep ranking
    
    return df

In [106]:
F200 = update_list(F100, 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1955/101.html')

In [107]:
F300 = update_list(F200, 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1955/201.html')

In [108]:
F400 = update_list(F300, 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1955/301.html')

In [109]:
F500 = update_list(F400, 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1955/401.html')

In [121]:
F500_55 = F500

## Link Pipeline - Loop

In [115]:
F500_56_links = [
    'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/',
    'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/101.html',
    'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/201.html',
    'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/301.html',
    'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/401.html'
]

In [116]:
F500_56 = pd.DataFrame(columns=['Company Name'])

In [118]:
for url in F500_56_links:
    F500_56 = update_list(F500_56, url)

In [120]:
F500_56.shape

(500, 1)

## Link Pipeline - Automated

In [122]:
print('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/' + '101' + '.html')

http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/101.html


In [126]:
for i in [101, 201, 301, 401]:
    print('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/' + str(i) + '.html')

http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/101.html
http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/201.html
http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/301.html
http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/401.html


In [127]:
F500_57_links = []

In [128]:
F500_57_links.append('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/')

In [129]:
F500_57_links

['http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/']

In [130]:
for i in [101, 201, 301, 401]:
    F500_57_links.append('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/' + str(i) + '.html')

In [131]:
F500_57_links

['http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/101.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/201.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/301.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/401.html']

In [132]:
F500_57 = pd.DataFrame(columns=['Company Name'])

for url in F500_57_links:
    F500_57 = update_list(F500_57, url)

F500_57.shape

(500, 1)

In [135]:
assert F500_57.shape == (500,1)

## Year Pipeline

In [136]:
year = 1958

In [137]:
F500_58_links = []

In [139]:
F500_58_links.append('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' + str(year) + '/')

In [142]:
for i in [101, 201, 301, 401]:
    F500_58_links.append('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' + str(year) + '/' + str(i) + '.html')

In [143]:
F500_58_links

['http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958/',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958/101.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958/201.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958/301.html',
 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958/401.html']

In [144]:
F500_58 = pd.DataFrame(columns=['Company Name'])

for url in F500_58_links:
    F500_58 = update_list(F500_58, url)

assert F500_58.shape == (500,1)

## Year Pipeline Automated

In [146]:
link_stem = 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' # all links build from this one

In [148]:
print(link_stem + str(year))

http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1958


In [149]:
len(F500_58_links)

5

In [157]:
def get_F500(year):
    """
    Creates a DataFrame with the names of the Fortune 500 companies from a given year
    Creates a list of urls with some gimmicky string manipulation, then extracts/transforms with the update_list function
    Returns a DataFrame with the list ordered by Revenue, the website default
    """
    links = [] # empty links table to be populated
    
    link_stem = 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' # all links build from this one
    
    links.append(link_stem + str(year)) # adds link to the first page in the list
    
    for i in [101, 201, 301, 401]:
        links.append(link_stem + str(year) + '/' + str(i) + '.html') # adds link to the following pages
    
    assert len(links) == 5
    
    df = pd.DataFrame(columns=['Company Name']) # instantiate empty DataFrame

    for url in links:
        df = update_list(df, url)    

    assert df.shape == (500,1)
    
    return df

In [158]:
F500_59 = get_F500(1959)

In [160]:
get_F500(1960)

Company Name
0                  General Motors
1                     Exxon Mobil
2                      Ford Motor
3                General Electric
4                      U.S. Steel
5                           Mobil
6                        Gulf Oil
7                          Texaco
8                        Chrysler
9                          Esmark
10              AT&T Technologies
11                         DuPont
12                Bethlehem Steel
13                          Amoco
14                            CBS
15                         Armour
16               General Dynamics
17                      Shell Oil
18                         Boeing
19                          Kraft
20         Goodyear Tire & Rubber
21                  ChevronTexaco
22                  Union Carbide
23                            RCA
24               Procter & Gamble
25         Navistar International
26        Intl. Business Machines
27                Lockheed Martin
28                   Sinclair Oil
29        Firestone Tire & Rubber
..                            ...
470             Permanente Cement
471                  Lear Siegler
472               Tennessee Corp.
473            American Petrofina
474                      Daystrom
475                          Hupp
476                 Tobin Packing
477                       Celotex
478              Columbian Carbon
479            Olivetti Underwood
480                  Bucyrus-Erie
481                Gardner-Denver
482                Remington Arms
483                  Bliss (E.W.)
484                        McCall
485            Bibb Manufacturing
486                        Friden
487                        Tappan
488  Standard Kollsman Industries
489                 Ekco Products
490  Carey (Philip) Manufacturing
491                      Walworth
492         Anderson-Prichard Oil
493                  Oxford Paper
494             Tung-Sol Electric
495               Universal Match
496            Miles Laboratories
497       Champlin Oil & Refining
498               Bigelow-Sanford
499                      Masonite

[500 rows x 1 columns]

# All Years DataFrame

In [195]:
year = 1955

In [180]:
df_55 = get_F500(1955)

In [181]:
df_56 = get_F500(1956)

In [185]:
df_5556 = pd.concat([df_55, df_56],axis=1)

In [201]:
df_5556 = pd.concat([get_F500(1955), get_F500(1956)], axis=1)

In [ ]:
base_df = get_F500(1955)
for i in range(1956, 1960):
    base_df = pd.concat([base_df, get_F500(i)], axis=1)

In [209]:
base_df.shape

(500, 50)

# All Years Improved

The problem with the previous All Years solution is that the column names don't reflect useful information (like the year) - and rather than renaming the columns, I'm gonna rewrite the initial functions to provide years in the column name

In [ ]:
def update_list(df, url):
    r = requests.get(url) # gets all the data from a webpage
    assert r.status_code == requests.codes.ok # checks that we got something
    soup = BeautifulSoup(r.text)
    
    entries = [] # list of topics from the webpage, includes some extraneous info
    for wrapper in soup.find_all('a'):
        entries.append(wrapper.text)
    
    companies = entries[5:-11] # slices the irrelevant information
    assert len(companies) == 100 # should only have 100 companies
    
    F5 = pd.DataFrame(data = companies, columns = df.columns) # create a dataframe from the list to concatenate
    assert df.columns == F5.columns # should be identical because reference above
    assert df.shape[1] == F5.shape[1] # verifies they have the same number of columns
    
    df = pd.concat([df, F5], 
                   ignore_index=True) # ignore index to keep ranking
    
    return df

So this is kind of a weird spot - the column names are coming from the 'update_list' function, which tells pandas to use the column names in the original dataframe -- I think I have to wholesale rewrite this to include the year from the url -- gonna look ugly

In [210]:
url = 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1956/101.html'

In [219]:
len(url.split('/'))

9

In [218]:
for i in url.split('/'):
    print(i)

http:

archive.fortune.com
magazines
fortune
fortune500_archive
full
1956
101.html


In [222]:
chunks = []
for i in url.split('/'):
    chunks.append(i)

In [225]:
chunks[-2]

'1956'

In [226]:
int(chunks[-2])

1956

In [236]:
def get_year(url):
    """
    Slices the url to find the year - returns it as a string for use in 'update_list_year' function
    """
    chunks = [] # empty list to be filled with the split pieces from url.split()
    
    for i in url.split('/'):
        chunks.append(i) # splits the url and appends each chunk to the chunks list
    
    if len(chunks) == 8: # the first page only have 8 chunks, omitting 101.html or similar
        year = int(chunks[-1])
    
    elif len(chunks) == 9: # the remaining pages have 9 chunks, because of 101.html etc.
        year = int(chunks[-2])
    
    return str(year)

In [237]:
get_year('http://archive.fortune.com/magazines/fortune/fortune500_archive/full/1957/101.html')

'1957'

In [243]:
def get_F500(year):
    """
    Creates a DataFrame with the names of the Fortune 500 companies from a given year
    Creates a list of urls with some gimmicky string manipulation, then extracts/transforms with the update_list function
    Returns a DataFrame with the list ordered by Revenue, the website default
    """
    links = [] # empty links table to be populated
    
    link_stem = 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' # all links build from this one
    
    links.append(link_stem + str(year)) # adds link to the first page in the list
    
    for i in [101, 201, 301, 401]:
        links.append(link_stem + str(year) + '/' + str(i) + '.html') # adds link to the following pages
    
    assert len(links) == 5
    
    df = pd.DataFrame(columns=[str(year)]) # instantiate empty DataFrame

    for url in links:
        df = update_list(df, url)    

    assert df.shape == (500,1)
    
    return df

In [248]:
base_df = get_F500(1955)
for i in range(1956, 2006):
    base_df = pd.concat([base_df, get_F500(i)], axis=1)

In [256]:
base_df.index -= 1

In [257]:
base_df['rank'] = base_df.index + 1

In [259]:
base_df.head()

1955            1956              1957              1958  \
0    General Motors  General Motors    General Motors    General Motors   
1       Exxon Mobil     Exxon Mobil       Exxon Mobil       Exxon Mobil   
2        U.S. Steel      Ford Motor        Ford Motor        Ford Motor   
3  General Electric      U.S. Steel        U.S. Steel        U.S. Steel   
4            Esmark        Chrysler  General Electric  General Electric   

               1959              1960              1961              1962  \
0    General Motors    General Motors    General Motors    General Motors   
1       Exxon Mobil       Exxon Mobil       Exxon Mobil       Exxon Mobil   
2        Ford Motor        Ford Motor        Ford Motor        Ford Motor   
3  General Electric  General Electric  General Electric  General Electric   
4        U.S. Steel        U.S. Steel        U.S. Steel             Mobil   

               1963              1964 ...               1997  \
0    General Motors    General Motors ...     General Motors   
1       Exxon Mobil       Exxon Mobil ...         Ford Motor   
2        Ford Motor        Ford Motor ...        Exxon Mobil   
3  General Electric  General Electric ...    Wal-Mart Stores   
4             Mobil             Mobil ...   General Electric   

               1998              1999              2000              2001  \
0    General Motors    General Motors    General Motors       Exxon Mobil   
1        Ford Motor        Ford Motor   Wal-Mart Stores   Wal-Mart Stores   
2       Exxon Mobil   Wal-Mart Stores       Exxon Mobil    General Motors   
3   Wal-Mart Stores       Exxon Mobil        Ford Motor        Ford Motor   
4  General Electric  General Electric  General Electric  General Electric   

              2002              2003              2004              2005 rank  
0  Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores    1  
1      Exxon Mobil    General Motors       Exxon Mobil       Exxon Mobil    2  
2   General Motors       Exxon Mobil    General Motors    General Motors    3  
3       Ford Motor        Ford Motor        Ford Motor        Ford Motor    4  
4            Enron  General Electric  General Electric  General Electric    5  

[5 rows x 52 columns]

In [260]:
cols = base_df.columns.tolist()

In [262]:
cols = cols[-1:] + cols[:-1]

In [264]:
data = base_df[cols]

In [266]:
data.head()

rank              1955            1956              1957              1958  \
0     1    General Motors  General Motors    General Motors    General Motors   
1     2       Exxon Mobil     Exxon Mobil       Exxon Mobil       Exxon Mobil   
2     3        U.S. Steel      Ford Motor        Ford Motor        Ford Motor   
3     4  General Electric      U.S. Steel        U.S. Steel        U.S. Steel   
4     5            Esmark        Chrysler  General Electric  General Electric   

               1959              1960              1961              1962  \
0    General Motors    General Motors    General Motors    General Motors   
1       Exxon Mobil       Exxon Mobil       Exxon Mobil       Exxon Mobil   
2        Ford Motor        Ford Motor        Ford Motor        Ford Motor   
3  General Electric  General Electric  General Electric  General Electric   
4        U.S. Steel        U.S. Steel        U.S. Steel             Mobil   

               1963        ...                    1996              1997  \
0    General Motors        ...          General Motors    General Motors   
1       Exxon Mobil        ...              Ford Motor        Ford Motor   
2        Ford Motor        ...             Exxon Mobil       Exxon Mobil   
3  General Electric        ...         Wal-Mart Stores   Wal-Mart Stores   
4             Mobil        ...                    AT&T  General Electric   

               1998              1999              2000              2001  \
0    General Motors    General Motors    General Motors       Exxon Mobil   
1        Ford Motor        Ford Motor   Wal-Mart Stores   Wal-Mart Stores   
2       Exxon Mobil   Wal-Mart Stores       Exxon Mobil    General Motors   
3   Wal-Mart Stores       Exxon Mobil        Ford Motor        Ford Motor   
4  General Electric  General Electric  General Electric  General Electric   

              2002              2003              2004              2005  
0  Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores  
1      Exxon Mobil    General Motors       Exxon Mobil       Exxon Mobil  
2   General Motors       Exxon Mobil    General Motors    General Motors  
3       Ford Motor        Ford Motor        Ford Motor        Ford Motor  
4            Enron  General Electric  General Electric  General Electric  

[5 rows x 52 columns]

In [267]:
data.to_csv('Fortune_500_1955_2005.csv')

# Cleaned + Pressed

In [250]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

In [251]:
def update_list(df, url):
    """
    Retrieves information from a url and concatenates it with an existing DataFrame
    
    Limitations:
         -slicing assumes Fortune Archives (http://archive.fortune.com/magazines/fortune/fortune500_archive/full/)
         -requires a pre-existing DataFrame with which to concatenate
         
    The second limitation is mitigated by the get_F500 function:
         -get_F500 sets the column name to the year embedded in the url
    """

    r = requests.get(url) # gets all the data from a webpage
    assert r.status_code == requests.codes.ok # checks that we got something
    
    soup = BeautifulSoup(r.text)
    
    entries = [] # list of topics from the webpage, includes some extraneous info
    for wrapper in soup.find_all('a'):
        entries.append(wrapper.text)
    
    companies = entries[5:-11] # slices the irrelevant information
    assert len(companies) == 100 # should only have 100 companies
    
    F5 = pd.DataFrame(data = companies, columns = df.columns) # create a dataframe from the list to concatenate
    assert df.columns == F5.columns # should be identical because they're defined that way
    assert df.shape[1] == F5.shape[1] # verifies they have the same number of columns
    
    df = pd.concat([df, F5], 
                   ignore_index=True) # ignore index to keep ranking
    
    return df

In [ ]:
def get_F500(year):
    """
    Creates a DataFrame with the names of the Fortune 500 companies from a given year
    Creates a list of urls with some gimmicky string manipulation, then extracts/transforms with the update_list function
    Returns a DataFrame with the list ordered by Revenue, the website default
    """
    links = [] # empty links table to be populated
    
    link_stem = 'http://archive.fortune.com/magazines/fortune/fortune500_archive/full/' # all links build from this one
    
    links.append(link_stem + str(year)) # adds link to the first page in the list
    
    for i in [101, 201, 301, 401]:
        links.append(link_stem + str(year) + '/' + str(i) + '.html') # adds link to the following pages
    
    assert len(links) == 5
    
    df = pd.DataFrame(columns=[str(year)]) # instantiate empty DataFrame

    for url in links:
        df = update_list(df, url)    

    assert df.shape == (500,1)
    
    return df

In [ ]:
base_df = get_F500(1955)
for i in range(1956, 2006):
    base_df = pd.concat([base_df, get_F500(i)], axis=1)

# Awkward Pivot 

I think years on the index and companies as columns would be the most useful for analysis, the problem is that pivoting my DataFrame isn't an obvious solution -- I'll try getting a list of companies first

In [354]:
Company_List = pd.unique(base_df.values.ravel())

In [459]:
Company_List[:10]

array(['General Motors', 'Exxon Mobil', 'Wal-Mart Stores', 1,
       'Ford Motor', 'Mobil', 2, 'U.S. Steel', 'Texaco', 3], dtype=object)

Unfortunately, the company list also includes the numbers 1 to 500 - those will get removed later

In [356]:
import numpy as np

In [357]:
index_list = np.arange(1955, 2006) # reflects the number of years for which I have data

In [358]:
df = pd.DataFrame(columns = Company_List, index=index_list)

In [361]:
df.shape # 51 years, 1887 companies -- still need to remove those numeric columns

(51, 2387)

In [457]:
data.head()

rank              1955            1956              1957              1958  \
0     1    General Motors  General Motors    General Motors    General Motors   
1     2       Exxon Mobil     Exxon Mobil       Exxon Mobil       Exxon Mobil   
2     3        U.S. Steel      Ford Motor        Ford Motor        Ford Motor   
3     4  General Electric      U.S. Steel        U.S. Steel        U.S. Steel   
4     5            Esmark        Chrysler  General Electric  General Electric   

               1959              1960              1961              1962  \
0    General Motors    General Motors    General Motors    General Motors   
1       Exxon Mobil       Exxon Mobil       Exxon Mobil       Exxon Mobil   
2        Ford Motor        Ford Motor        Ford Motor        Ford Motor   
3  General Electric  General Electric  General Electric  General Electric   
4        U.S. Steel        U.S. Steel        U.S. Steel             Mobil   

               1963        ...                    1996              1997  \
0    General Motors        ...          General Motors    General Motors   
1       Exxon Mobil        ...              Ford Motor        Ford Motor   
2        Ford Motor        ...             Exxon Mobil       Exxon Mobil   
3  General Electric        ...         Wal-Mart Stores   Wal-Mart Stores   
4             Mobil        ...                    AT&T  General Electric   

               1998              1999              2000              2001  \
0    General Motors    General Motors    General Motors       Exxon Mobil   
1        Ford Motor        Ford Motor   Wal-Mart Stores   Wal-Mart Stores   
2       Exxon Mobil   Wal-Mart Stores       Exxon Mobil    General Motors   
3   Wal-Mart Stores       Exxon Mobil        Ford Motor        Ford Motor   
4  General Electric  General Electric  General Electric  General Electric   

              2002              2003              2004              2005  
0  Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores   Wal-Mart Stores  
1      Exxon Mobil    General Motors       Exxon Mobil       Exxon Mobil  
2   General Motors       Exxon Mobil    General Motors    General Motors  
3       Ford Motor        Ford Motor        Ford Motor        Ford Motor  
4            Enron  General Electric  General Electric  General Electric  

[5 rows x 52 columns]

In [365]:
data.index[data['1955'] == 'General Motors'].tolist() # means that GM was the top ranked Fortunate 500 company in 1955

[0]

## Get Indices for One Company

I want a list of yearly indices for each company, which I can then add to the empty DataFrame above

In [463]:
# GM example because it's easy to verify

GM_indices = []
for i in range(1955,2006):
    idx = data.index[data[str(i)] == 'General Motors'].tolist()
    GM_indices.append(idx)

In [382]:
type(GM_indices)

list

In [375]:
#Target example because I want to see how my approach handles null values

Target_indices = []
for i in range(1955,2006):
    idx = data.index[data[str(i)] == 'Target'].tolist()
    Target_indices.append(idx)

In [464]:
df['General Motors'] = GM_indices

In [465]:
df.head()

General Motors Exxon Mobil Wal-Mart Stores   1 Ford Motor Mobil   2  \
1955            [0]         [1]              []  []         []   [8]  []   
1956            [0]         [1]              []  []        [2]  [12]  []   
1957            [0]         [1]              []  []        [2]   [5]  []   
1958            [0]         [1]              []  []        [2]   [6]  []   
1959            [0]         [1]              []  []        [2]   [5]  []   

     U.S. Steel Texaco   3 ... Weatherhead 499 Draper Itek  \
1955        [2]   [13]  [] ...          []  []     []   []   
1956        [3]   [14]  [] ...       [498]  []     []   []   
1957        [3]   [11]  [] ...          []  []     []   []   
1958        [3]   [11]  [] ...          []  []     []   []   
1959        [4]    [8]  [] ...          []  []     []   []   

     Winnebago Industries Rochester & Pittsburgh Coal ReliaStar Financial  \
1955                   []                          []                  []   
1956                   []                          []                  []   
1957                   []                          []                  []   
1958                   []                          []                  []   
1959                   []                          []                  []   

     Neiman Marcus Cincinnati Financial 500  
1955            []                   []  []  
1956            []                   []  []  
1957            []                   []  []  
1958            []                   []  []  
1959            []                   []  []  

[5 rows x 2387 columns]

So, the rank becomes 0 instead of 1, I can live with that

## Index Getting Function

In [439]:
def get_idx(company, df):
    indices = []
    for i in range(1955,2006):
        idx = df.index[df[str(i)] == company].tolist()
        indices.append(idx)
    return indices

In [467]:
# Test with the Nieman Marcus data - works, returns a list

NM_idx = get_idx('Neiman Marcus', data)

type(NM_idx)

list

In [470]:
# Made a copy of the Company List and df for convenience and backup
CL = Company_List


df_test = df

In [471]:
# Exxon Mobil test was successful

EX_idx = get_idx('Exxon Mobil', data)

df_test['Exxon Mobil'] = EX_idx

df_test['Exxon Mobil'].head()

1955    [1]
1956    [1]
1957    [1]
1958    [1]
1959    [1]
Name: Exxon Mobil, dtype: object

In [445]:
# Now to run it across all of the companies

for company in CL:
    idx = get_idx(company, data)
    df_test[company] = idx

In [472]:
df_test.head()

General Motors Exxon Mobil Wal-Mart Stores   1 Ford Motor Mobil   2  \
1955            [0]         [1]              []  []         []   [8]  []   
1956            [0]         [1]              []  []        [2]  [12]  []   
1957            [0]         [1]              []  []        [2]   [5]  []   
1958            [0]         [1]              []  []        [2]   [6]  []   
1959            [0]         [1]              []  []        [2]   [5]  []   

     U.S. Steel Texaco   3 ... Weatherhead 499 Draper Itek  \
1955        [2]   [13]  [] ...          []  []     []   []   
1956        [3]   [14]  [] ...       [498]  []     []   []   
1957        [3]   [11]  [] ...          []  []     []   []   
1958        [3]   [11]  [] ...          []  []     []   []   
1959        [4]    [8]  [] ...          []  []     []   []   

     Winnebago Industries Rochester & Pittsburgh Coal ReliaStar Financial  \
1955                   []                          []                  []   
1956                   []                          []                  []   
1957                   []                          []                  []   
1958                   []                          []                  []   
1959                   []                          []                  []   

     Neiman Marcus Cincinnati Financial 500  
1955            []                   []  []  
1956            []                   []  []  
1957            []                   []  []  
1958            []                   []  []  
1959            []                   []  []  

[5 rows x 2387 columns]

Save to a CSV - will remove the brackets with regular expressions via gnumeric

In [447]:
df_test.to_csv('Fortune_500_ranks.csv')

Loading the bracket-less data

In [449]:
df_cleaned = pd.read_csv('Fortune_500_ranks_cleaned.csv')

In [451]:
df_cleaned.head() # We can still see the numeric columns in the data - fortunately those are entirely Null so...

Unnamed: 0  General Motors  Exxon Mobil  Wal-Mart Stores   1  Ford Motor  \
0        1955               0            1              NaN NaN         NaN   
1        1956               0            1              NaN NaN         2.0   
2        1957               0            1              NaN NaN         2.0   
3        1958               0            1              NaN NaN         2.0   
4        1959               0            1              NaN NaN         2.0   

   Mobil   2  U.S. Steel  Texaco ...   Weatherhead  499  Draper  Itek  \
0    8.0 NaN         2.0    13.0 ...           NaN  NaN     NaN   NaN   
1   12.0 NaN         3.0    14.0 ...         498.0  NaN     NaN   NaN   
2    5.0 NaN         3.0    11.0 ...           NaN  NaN     NaN   NaN   
3    6.0 NaN         3.0    11.0 ...           NaN  NaN     NaN   NaN   
4    5.0 NaN         4.0     8.0 ...           NaN  NaN     NaN   NaN   

   Winnebago Industries  Rochester & Pittsburgh Coal  ReliaStar Financial  \
0                   NaN                          NaN                  NaN   
1                   NaN                          NaN                  NaN   
2                   NaN                          NaN                  NaN   
3                   NaN                          NaN                  NaN   
4                   NaN                          NaN                  NaN   

   Neiman Marcus  Cincinnati Financial  500  
0            NaN                   NaN  NaN  
1            NaN                   NaN  NaN  
2            NaN                   NaN  NaN  
3            NaN                   NaN  NaN  
4            NaN                   NaN  NaN  

[5 rows x 2388 columns]

In [453]:
df_cleaned.dropna(how='all', axis=1, inplace=True) # Dropped

In [454]:
df_cleaned.shape

(51, 1888)

Saved to a csv in order to use Tableau

In [456]:
df_cleaned.to_csv('Fortune_500_ranked_by_year.csv')

In [493]:
df_cleaned['CBS'][-15:-5]

36            32
37            30
38            34
39            41
40    [120, 305]
41           134
42           155
43           158
44           171
45           236
Name: CBS, dtype: object

In [487]:
data['1995'][305]

'CBS'

Found an abberration - some companies appear twice in the Fortune500 Rankings, for example CBS appears at 120th and 305th in the year 1995

## Pivot

If I want to create a multi-line lineplot in tableau, I need to pivot the DataFrame so that each company is an entry under a 'Company' column, while the Years becomes the column name again

In [494]:
df_cleaned.head()

Unnamed: 0  General Motors  Exxon Mobil  Wal-Mart Stores  Ford Motor  \
0        1955               0            1              NaN         NaN   
1        1956               0            1              NaN         2.0   
2        1957               0            1              NaN         2.0   
3        1958               0            1              NaN         2.0   
4        1959               0            1              NaN         2.0   

   Mobil  U.S. Steel  Texaco  General Electric  Chrysler  \
0    8.0         2.0    13.0                 3       5.0   
1   12.0         3.0    14.0                 5       4.0   
2    5.0         3.0    11.0                 4       6.0   
3    6.0         3.0    11.0                 4       5.0   
4    5.0         4.0     8.0                 3      10.0   

           ...           Wendy's International  Goodall-Sanford  Weatherhead  \
0          ...                             NaN            498.0          NaN   
1          ...                             NaN              NaN        498.0   
2          ...                             NaN              NaN          NaN   
3          ...                             NaN              NaN          NaN   
4          ...                             NaN              NaN          NaN   

   Draper  Itek  Winnebago Industries  Rochester & Pittsburgh Coal  \
0     NaN   NaN                   NaN                          NaN   
1     NaN   NaN                   NaN                          NaN   
2     NaN   NaN                   NaN                          NaN   
3     NaN   NaN                   NaN                          NaN   
4     NaN   NaN                   NaN                          NaN   

   ReliaStar Financial  Neiman Marcus  Cincinnati Financial  
0                  NaN            NaN                   NaN  
1                  NaN            NaN                   NaN  
2                  NaN            NaN                   NaN  
3                  NaN            NaN                   NaN  
4                  NaN            NaN                   NaN  

[5 rows x 1888 columns]

In [512]:
df_transposed = df_cleaned.transpose()

In [513]:
df_transposed.head(5)

0     1     2     3     4     5     6     7     8     9   \
Unnamed: 0       1955  1956  1957  1958  1959  1960  1961  1962  1963  1964   
General Motors      0     0     0     0     0     0     0     0     0     0   
Exxon Mobil         1     1     1     1     1     1     1     1     1     1   
Wal-Mart Stores   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Ford Motor        NaN     2     2     2     2     2     2     2     2     2   

                 ...     41    42    43    44    45    46    47    48    49  \
Unnamed: 0       ...   1996  1997  1998  1999  2000  2001  2002  2003  2004   
General Motors   ...      0     0     0     0     0     2     2     1     2   
Exxon Mobil      ...      2     2     2     3     2     0     1     2     1   
Wal-Mart Stores  ...      3     3     3     2     1     1     0     0     0   
Ford Motor       ...      1     1     1     1     3     3     3     3     3   

                   50  
Unnamed: 0       2005  
General Motors      2  
Exxon Mobil         1  
Wal-Mart Stores     0  
Ford Motor          3  

[5 rows x 51 columns]

In [514]:
df_transposed.columns = df_transposed.iloc[0]

In [515]:
df_transposed.head(5)

Unnamed: 0       1955  1956  1957  1958  1959  1960  1961  1962  1963  1964  \
Unnamed: 0       1955  1956  1957  1958  1959  1960  1961  1962  1963  1964   
General Motors      0     0     0     0     0     0     0     0     0     0   
Exxon Mobil         1     1     1     1     1     1     1     1     1     1   
Wal-Mart Stores   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Ford Motor        NaN     2     2     2     2     2     2     2     2     2   

Unnamed: 0       ...   1996  1997  1998  1999  2000  2001  2002  2003  2004  \
Unnamed: 0       ...   1996  1997  1998  1999  2000  2001  2002  2003  2004   
General Motors   ...      0     0     0     0     0     2     2     1     2   
Exxon Mobil      ...      2     2     2     3     2     0     1     2     1   
Wal-Mart Stores  ...      3     3     3     2     1     1     0     0     0   
Ford Motor       ...      1     1     1     1     3     3     3     3     3   

Unnamed: 0       2005  
Unnamed: 0       2005  
General Motors      2  
Exxon Mobil         1  
Wal-Mart Stores     0  
Ford Motor          3  

[5 rows x 51 columns]

In [520]:
df_transposed = df_transposed.drop(df_transposed.index[0])

In [525]:
df_transposed.index.names = ['Company']

In [529]:
df_transposed.head(5)

Unnamed: 0      1955 1956 1957 1958 1959 1960 1961 1962 1963 1964 ...  1996  \
Company                                                           ...         
General Motors     0    0    0    0    0    0    0    0    0    0 ...     0   
Exxon Mobil        1    1    1    1    1    1    1    1    1    1 ...     2   
Wal-Mart Stores  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...     3   
Ford Motor       NaN    2    2    2    2    2    2    2    2    2 ...     1   
Mobil              8   12    5    6    5    5    5    4    4    4 ...     7   

Unnamed: 0      1997 1998 1999 2000 2001 2002 2003 2004 2005  
Company                                                       
General Motors     0    0    0    0    2    2    1    2    2  
Exxon Mobil        2    2    3    2    0    1    2    1    1  
Wal-Mart Stores    3    3    2    1    1    0    0    0    0  
Ford Motor         1    1    1    3    3    3    3    3    3  
Mobil              7    7   12  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 51 columns]

In [530]:
df_transposed.to_csv('Fortune_500_rank_Company_Indexed.csv')

# To Do

I realize what I need to do:  
I need to create a table with companies in the first column, years in the second column - it'll be a friggen enormous table because it will have (# companies * # years) rows -- the third column will be the companies rank in the Fortune 500  in that year 

In [21]:
import pandas as pd
import numpy as np

In [22]:
data = pd.read_csv('Fortune_500_rank_Company_Indexed.csv')

In [23]:
data.head()

Company  1955  1956  1957  1958  1959  1960  1961  1962  1963 ...   \
0   General Motors   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0 ...    
1      Exxon Mobil   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0 ...    
2  Wal-Mart Stores   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN ...    
3       Ford Motor   NaN   2.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0 ...    
4            Mobil   8.0  12.0   5.0   6.0   5.0   5.0   5.0   4.0   4.0 ...    

   1996  1997  1998  1999  2000  2001 2002 2003 2004 2005  
0     0     0     0     0     0     2    2    1  2.0  2.0  
1     2     2     2     3     2     0    1    2  1.0  1.0  
2   3.0   3.0   3.0   2.0   1.0   1.0  0.0  0.0  0.0  0.0  
3   1.0   1.0   1.0   1.0   3.0   3.0  3.0  3.0  3.0  3.0  
4   7.0   7.0   7.0  12.0   NaN   NaN  NaN  NaN  NaN  NaN  

[5 rows x 52 columns]

In [27]:
data.columns

Index(['Company', '1955', '1956', '1957', '1958', '1959', '1960', '1961',
       '1962', '1963', '1964', '1965', '1966', '1967', '1968', '1969', '1970',
       '1971', '1972', '1973', '1974', '1975', '1976', '1977', '1978', '1979',
       '1980', '1981', '1982', '1983', '1984', '1985', '1986', '1987', '1988',
       '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997',
       '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005'],
      dtype='object')

In [47]:
Fortune_500_unpivoted = pd.melt(data, id_vars=['Company'], value_vars=data.columns[1:])

In [48]:
Fortune_500_unpivoted.rename(columns={'Company':'Company',
                                     'variable': 'Year',
                                     'value': 'Rank'},
                            inplace=True)

In [33]:
Fortune_500_unpivoted.to_csv('Fortune_500_unpivoted.csv')

## Prep for Log

The current data is all well and good, but if I want to take the log of rank I need to add 1 to every rank in order to register the top ranked companies (log of 0 disappears)

In [49]:
F500_unp = Fortune_500_unpivoted

In [52]:
F500_cln = F500_unp.dropna()

In [76]:
for i in F500_cln['Rank'].values:
    if type(i) != float:
        print(len(i))
        print(i)

1
0
1
1
3
2.0
3
6.0
4
11.0
3
7.0
1
3
3
5.0
1
4
4
21.0
2
12
4
39.0
3
9.0
4
17.0
2
14
3
144
2
37
4
24.0
3
8.0
4
10.0
2
25
4
28.0
5
292.0
4
15.0
2
27
2
16
2
32
4
22.0
4
13.0
4
34.0
3
125
4
43.0
4
61.0
4
20.0
2
35
3
118
4
33.0
5
434.0
4
31.0
4
18.0
2
46
3
173
2
19
2
30
2
26
2
48
2
40
2
29
4
68.0
4
23.0
4
96.0
4
70.0
5
103.0
3
131
4
36.0
2
41
3
138
2
80
4
82.0
2
57
4
67.0
5
178.0
4
65.0
2
58
4
64.0
3
169
4
45.0
4
50.0
2
51
5
115.0
2
78
4
53.0
2
66
4
55.0
4
42.0
3
127
4
59.0
2
97
4
38.0
3
100
3
148
2
88
4
89.0
5
151.0
4
62.0
3
122
4
44.0
3
168
4
52.0
4
49.0
5
256.0
3
294
4
75.0
4
71.0
5
325.0
2
56
2
79
4
47.0
2
85
4
92.0
5
146.0
5
105.0
4
74.0
4
63.0
5
156.0
4
77.0
5
139.0
4
90.0
3
102
4
73.0
2
54
3
192
3
129
3
101
4
60.0
5
317.0
4
91.0
4
76.0
5
134.0
4
98.0
5
107.0
3
247
5
143.0
5
108.0
3
111
2
93
5
188.0
5
110.0
4
72.0
5
176.0
5
273.0
5
238.0
5
121.0
4
95.0
5
152.0
4
87.0
4
69.0
5
106.0
5
187.0
4
86.0
2
99
5
119.0
2
84
5
198.0
3
133
3
130
5
113.0
2
81
5
166.0
5
128.0
3
141
5
150.0
5
126.0


5
208.0
5
180.0
5
203.0
5
417.0
5
300.0
5
198.0
5
189.0
5
409.0
5
336.0
5
254.0
5
207.0
5
434.0
5
431.0
5
278.0
5
271.0
10
[323, 433]
5
283.0
5
223.0
5
269.0
5
370.0
5
267.0
5
376.0
5
385.0
5
395.0
5
287.0
5
279.0
5
215.0
5
460.0
5
442.0
5
197.0
5
228.0
5
321.0
5
263.0
5
306.0
5
246.0
5
247.0
5
241.0
5
303.0
5
252.0
5
366.0
5
344.0
3
291
5
280.0
5
233.0
5
234.0
5
389.0
5
316.0
5
305.0
5
428.0
5
338.0
5
368.0
5
360.0
5
396.0
5
245.0
5
304.0
5
256.0
5
391.0
5
222.0
5
349.0
5
367.0
5
268.0
5
467.0
5
253.0
5
238.0
5
226.0
5
334.0
5
495.0
5
365.0
5
379.0
5
430.0
5
318.0
5
411.0
5
356.0
5
288.0
5
297.0
5
354.0
5
312.0
5
248.0
5
285.0
5
399.0
5
255.0
5
275.0
5
307.0
5
386.0
5
299.0
5
343.0
5
310.0
5
342.0
5
347.0
5
448.0
5
404.0
5
277.0
5
274.0
5
325.0
5
261.0
5
293.0
5
497.0
5
296.0
5
308.0
5
309.0
5
452.0
5
402.0
5
332.0
5
284.0
5
352.0
5
463.0
5
292.0
5
478.0
5
414.0
5
480.0
5
380.0
5
276.0
5
421.0
5
444.0
5
282.0
5
374.0
5
488.0
5
499.0
5
372.0
5
328.0
5
394.0
5
315.0
5
432.0
5
329.0
5
48

5
204.0
5
485.0
5
392.0
5
259.0
5
462.0
5
175.0
5
428.0
5
212.0
5
344.0
5
267.0
5
252.0
5
404.0
5
351.0
10
[317, 458]
5
337.0
5
273.0
5
248.0
5
244.0
5
432.0
5
340.0
5
245.0
5
294.0
5
461.0
5
210.0
5
366.0
5
369.0
5
250.0
5
465.0
5
364.0
5
251.0
5
473.0
5
213.0
5
292.0
5
203.0
5
241.0
5
238.0
5
338.0
5
313.0
5
325.0
5
311.0
5
433.0
3
322
5
264.0
5
249.0
5
335.0
5
302.0
5
265.0
5
447.0
5
347.0
5
435.0
5
321.0
5
393.0
5
318.0
5
444.0
5
470.0
5
329.0
5
310.0
5
363.0
5
227.0
5
320.0
5
386.0
5
287.0
5
281.0
5
298.0
5
330.0
5
431.0
5
399.0
5
406.0
5
319.0
5
438.0
5
290.0
5
424.0
5
297.0
5
333.0
5
336.0
5
352.0
5
288.0
5
306.0
5
271.0
5
474.0
5
420.0
5
354.0
5
283.0
5
328.0
5
385.0
5
261.0
5
332.0
5
266.0
5
380.0
5
456.0
5
362.0
5
280.0
5
479.0
5
422.0
5
334.0
5
478.0
5
371.0
5
289.0
5
268.0
5
472.0
5
377.0
5
375.0
5
296.0
5
256.0
5
270.0
5
397.0
5
349.0
5
278.0
5
395.0
5
314.0
5
451.0
5
442.0
5
394.0
5
411.0
5
272.0
5
436.0
5
353.0
5
389.0
5
421.0
5
484.0
5
396.0
5
446.0
5
277.0
5
307.0
5
43

197.0
5
242.0
5
382.0
5
303.0
5
222.0
5
284.0
5
341.0
5
257.0
5
312.0
5
480.0
5
192.0
5
407.0
5
261.0
3
428
5
258.0
5
274.0
5
277.0
5
467.0
5
214.0
5
230.0
5
235.0
5
314.0
5
307.0
5
374.0
5
229.0
5
254.0
3
299
5
209.0
5
402.0
5
278.0
5
259.0
5
271.0
5
309.0
5
290.0
5
373.0
5
377.0
5
364.0
5
395.0
5
369.0
5
315.0
5
426.0
5
327.0
5
328.0
5
292.0
5
340.0
5
343.0
5
461.0
5
394.0
5
479.0
5
459.0
5
415.0
5
318.0
5
484.0
5
282.0
5
323.0
5
321.0
5
308.0
5
301.0
5
291.0
5
326.0
5
275.0
5
436.0
5
465.0
5
322.0
5
389.0
5
475.0
5
457.0
5
432.0
5
413.0
5
367.0
5
255.0
5
352.0
5
310.0
5
351.0
5
244.0
5
462.0
5
296.0
5
345.0
5
297.0
5
430.0
5
304.0
5
420.0
5
302.0
5
442.0
5
408.0
5
332.0
5
260.0
5
342.0
5
270.0
5
329.0
5
324.0
5
396.0
5
405.0
5
378.0
5
325.0
5
376.0
5
288.0
5
481.0
5
266.0
5
375.0
5
453.0
5
384.0
5
466.0
5
422.0
5
355.0
5
390.0
5
425.0
5
347.0
5
455.0
5
287.0
5
399.0
5
316.0
5
452.0
5
334.0
5
360.0
5
354.0
5
446.0
5
393.0
5
494.0
5
285.0
5
383.0
5
448.0
5
362.0
5
370.0
5
489.0
5
379.

412.0
5
468.0
5
309.0
5
201.0
5
319.0
5
248.0
5
410.0
5
492.0
5
181.0
5
346.0
5
278.0
5
244.0
5
286.0
5
166.0
5
297.0
5
214.0
5
302.0
5
381.0
5
374.0
5
417.0
5
402.0
5
173.0
5
276.0
5
198.0
5
332.0
5
423.0
5
223.0
5
364.0
5
478.0
5
445.0
5
324.0
5
359.0
5
341.0
5
393.0
5
441.0
5
416.0
5
211.0
5
385.0
5
321.0
5
209.0
5
426.0
5
180.0
5
457.0
5
327.0
5
479.0
5
448.0
5
480.0
5
376.0
5
474.0
5
188.0
5
466.0
5
199.0
5
189.0
5
455.0
5
329.0
5
203.0
5
460.0
5
358.0
5
210.0
5
225.0
5
392.0
5
268.0
5
213.0
5
338.0
5
404.0
3
471
5
369.0
5
399.0
5
357.0
5
438.0
5
361.0
5
234.0
5
472.0
5
379.0
5
204.0
5
235.0
5
206.0
3
422
5
280.0
5
443.0
5
212.0
5
256.0
5
215.0
5
424.0
5
242.0
5
330.0
5
251.0
5
464.0
5
386.0
5
310.0
5
413.0
5
224.0
5
247.0
5
226.0
5
461.0
5
262.0
5
458.0
5
228.0
5
447.0
5
260.0
5
241.0
5
395.0
5
257.0
5
342.0
5
246.0
5
250.0
5
418.0
5
367.0
5
487.0
5
282.0
5
481.0
5
267.0
5
469.0
5
499.0
5
331.0
5
333.0
5
435.0
5
498.0
5
266.0
5
352.0
5
450.0
5
328.0
5
273.0
5
315.0
5
375.0
5
293.

5
475.0
5
175.0
5
194.0
5
262.0
5
327.0
5
497.0
5
389.0
5
187.0
5
470.0
5
431.0
5
223.0
5
183.0
5
351.0
5
437.0
5
253.0
5
216.0
5
473.0
5
362.0
5
237.0
5
450.0
5
396.0
5
193.0
5
225.0
5
205.0
5
404.0
5
346.0
5
192.0
5
498.0
5
202.0
5
330.0
10
[358, 413]
5
406.0
5
266.0
5
198.0
5
247.0
5
386.0
5
217.0
5
200.0
5
257.0
5
487.0
5
490.0
5
317.0
5
229.0
5
265.0
3
400
5
208.0
5
215.0
5
462.0
5
491.0
5
489.0
5
220.0
5
259.0
5
418.0
5
233.0
5
221.0
5
260.0
5
427.0
5
222.0
5
298.0
5
458.0
5
381.0
5
224.0
5
394.0
5
316.0
5
493.0
5
441.0
5
318.0
5
275.0
5
306.0
5
398.0
5
308.0
5
339.0
5
250.0
5
380.0
5
295.0
5
429.0
5
329.0
5
270.0
5
315.0
5
459.0
5
479.0
5
300.0
5
258.0
5
342.0
5
297.0
5
263.0
5
348.0
5
269.0
5
399.0
5
336.0
5
272.0
5
359.0
5
425.0
5
360.0
5
312.0
5
376.0
5
388.0
5
495.0
5
435.0
5
350.0
5
414.0
5
280.0
5
482.0
5
287.0
5
283.0
5
456.0
5
319.0
5
416.0
5
378.0
5
289.0
5
472.0
5
443.0
5
293.0
5
453.0
5
294.0
5
314.0
5
321.0
5
420.0
5
393.0
5
344.0
5
338.0
5
477.0
5
430.0
5
461.0
5
47

Evidently there are some values of Rank which include two numbers, ex. [307, 312] - I'll clean that up with Regex in Gnumeric:  
find: $\[(\d*), (\d*)\]$  
replace: $1

In [77]:
F500_cln.to_csv('F500_cln.csv')

In [78]:
!ls

F500_cln.csv			      Fortune_500_ranks_cleaned.csv
F500_Extract.ipynb		      Fortune_500_ranks_cleaned.gnumeric
F500_log_prep.csv		      Fortune_500_ranks.csv
Fortune_500_1955_2005.csv	      Fortune_500_ranks_experiment.gnumeric
Fortune_500_rank_Company_Indexed.csv  Fortune_500_smooth_bump_chart.csv
Fortune_500_ranked_by_year.csv	      Fortune_500_unpivoted.csv


In [7]:
F500_log_prepped = pd.read_csv('F500_log_prep.csv')

In [8]:
F500_log_prepped.drop(['Unnamed: 0'], axis=1, inplace=True)

In [9]:
F500_log_prepped.head()

Company  Year  Rank
0  General Motors  1955     0
1     Exxon Mobil  1955     1
2           Mobil  1955     8
3      U.S. Steel  1955     2
4          Texaco  1955    13

In [98]:
F500_log_prepped['Rank'] = F500_log_prepped['Rank'].values.astype(int)

In [99]:
F500_log_prepped['Rank'] += 1

In [101]:
F500_log_prepped.to_csv('F500_log_prep_complete.csv')

In [1]:
import pandas as pd

In [2]:
query = pd.read_csv('F500_log_prep_complete.csv')

In [11]:
query['Unnamed: 0']

0            0
1            1
2            2
3            3
4            4
5            5
6            6
7            7
8            8
9            9
10          10
11          11
12          12
13          13
14          14
15          15
16          16
17          17
18          18
19          19
20          20
21          21
22          22
23          23
24          24
25          25
26          26
27          27
28          28
29          29
         ...  
25421    25421
25422    25422
25423    25423
25424    25424
25425    25425
25426    25426
25427    25427
25428    25428
25429    25429
25430    25430
25431    25431
25432    25432
25433    25433
25434    25434
25435    25435
25436    25436
25437    25437
25438    25438
25439    25439
25440    25440
25441    25441
25442    25442
25443    25443
25444    25444
25445    25445
25446    25446
25447    25447
25448    25448
25449    25449
25450    25450
Name: Unnamed: 0, Length: 25451, dtype: int64

## Smooth Bump Chart Prep

In [46]:
Fortune_500_unpivoted['Join'] = 1

In [42]:
Join = pd.read_excel('../Downloads/Join.xlsx')

In [ ]:
pd.merge(df_a, df_b, on='subject_id', how='outer')

In [44]:
Fortune_500_smooth_bump_chart = pd.merge(Fortune_500_unpivoted, Join, on='Join', how='outer')

In [45]:
Fortune_500_smooth_bump_chart.to_csv('Fortune_500_smooth_bump_chart.csv')